File: src/lib/github-auth.ts
import { invoke } from "@tauri-apps/api/core";
import type { AuthSession, DeviceFlowPending, AuthUser } from "@/types/auth";
import { normalizeAuthSession, normalizeAuthUser } from "@/lib/auth-session";
import { isTauri } from "@/lib/tauri";

const SCOPES = "read:user user:email repo";

const DEVICE_CODE_PATH = "/login/device/code";
const ACCESS_TOKEN_PATH = "/login/oauth/access_token";
const USER_API_PATH = "/user";

/** Dev-only Vite proxies avoid browser CORS when not running inside Tauri. */
const OAUTH_BASE = import.meta.env.DEV ? "/api/github-oauth" : "https://github.com";
const API_BASE = import.meta.env.DEV ? "/api/github" : "https://api.github.com";

export function getGitHubClientId(): string | undefined {
  const id = import.meta.env.VITE_GITHUB_CLIENT_ID;
  return id?.trim() || undefined;
}

export function isGitHubAuthConfigured(): boolean {
  return Boolean(getGitHubClientId());
}

type RawDeviceFlow = DeviceFlowPending & {
  device_code?: string;
  user_code?: string;
  verification_uri?: string;
  expires_in?: number;
};

export function normalizeDeviceFlowPending(raw: RawDeviceFlow): DeviceFlowPending {
  const deviceCode = raw.deviceCode ?? raw.device_code ?? "";
  const userCode = raw.userCode ?? raw.user_code ?? "";
  const verificationUri =
    raw.verificationUri ??
    raw.verification_uri ??
    "https://github.com/login/device";

  if (!deviceCode || !userCode) {
    throw new Error("GitHub returned an incomplete sign-in response. Please try again.");
  }

  return {
    deviceCode,
    userCode,
    verificationUri,
    expiresIn: Number(raw.expiresIn ?? raw.expires_in ?? 900),
    interval: Number(raw.interval ?? 5),
  };
}

async function postFormBrowser(
  url: string,
  body: Record<string, string>,
): Promise<Record<string, string>> {
  const res = await fetch(url, {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: new URLSearchParams(body).toString(),
  });

  const data = (await res.json()) as Record<string, string>;

  if (data.access_token) {
    return data;
  }

  const oauthError = data.error;
  if (
    oauthError === "authorization_pending" ||
    oauthError === "slow_down" ||
    oauthError === "expired_token" ||
    oauthError === "access_denied"
  ) {
    return data;
  }

  if (!res.ok) {
    throw new Error(data.error_description ?? data.error ?? "GitHub request failed");
  }

  return data;
}

export async function startGitHubDeviceFlow(): Promise<DeviceFlowPending> {
  const clientId = getGitHubClientId();
  if (!clientId) {
    throw new Error(
      "GitHub sign-in is not configured. Set VITE_GITHUB_CLIENT_ID in your .env file.",
    );
  }

  if (isTauri()) {
    const pending = await invoke<RawDeviceFlow>("github_start_device_flow", {
      clientId,
      scope: SCOPES,
    });
    return normalizeDeviceFlowPending(pending);
  }

  const data = await postFormBrowser(`${OAUTH_BASE}${DEVICE_CODE_PATH}`, {
    client_id: clientId,
    scope: SCOPES,
  });

  return normalizeDeviceFlowPending({
    deviceCode: data.device_code,
    userCode: data.user_code,
    verificationUri: data.verification_uri,
    expiresIn: Number(data.expires_in) || 900,
    interval: Number(data.interval) || 5,
  });
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function extractAccessTokenFromPoll(
  result: Record<string, unknown>,
): string | undefined {
  if (typeof result.accessToken === "string") return result.accessToken;
  if (typeof result.access_token === "string") return result.access_token;

  for (const key of ["Token", "token"]) {
    const nested = result[key];
    if (nested && typeof nested === "object") {
      const obj = nested as Record<string, unknown>;
      if (typeof obj.accessToken === "string") return obj.accessToken;
      if (typeof obj.access_token === "string") return obj.access_token;
    }
  }

  return undefined;
}

async function pollGitHubDeviceFlowTauri(
  pending: DeviceFlowPending,
  signal?: AbortSignal,
  onAttempt?: (attempt: number) => void,
): Promise<string> {
  const clientId = getGitHubClientId();
  if (!clientId) {
    throw new Error("GitHub client ID is not configured.");
  }

  const pendingWithExpiry = pending as DeviceFlowPending & { expiresAt?: number };
  const deadline =
    pendingWithExpiry.expiresAt ?? Date.now() + pending.expiresIn * 1000;
  let intervalMs = Math.max(pending.interval * 1000, 1000);
  let attempt = 0;

  while (Date.now() < deadline) {
    if (signal?.aborted) {
      throw new Error("Sign-in cancelled");
    }

    attempt += 1;
    onAttempt?.(attempt);

    let result: Record<string, unknown>;
    try {
      result = await invoke<Record<string, unknown>>("github_poll_device_token", {
        clientId,
        deviceCode: pending.deviceCode,
      });
    } catch (e) {
      const wrapped = new Error(
        e instanceof Error ? e.message : "Could not reach GitHub to complete sign-in.",
      );
      Object.assign(wrapped, { cause: e });
      throw wrapped;
    }

    const status = String(result.status ?? "").toLowerCase();
    const token = extractAccessTokenFromPoll(result);
    if (token) return token;

    if (status === "slowdown" || status === "slow_down") {
      intervalMs += 5000;
    } else if (status && status !== "pending") {
      throw new Error(
        typeof result.error === "string"
          ? result.error
          : `GitHub sign-in failed (${status}).`,
      );
    }

    await delay(intervalMs);
  }

  throw new Error("Sign-in timed out. Please try again.");
}

export async function pollGitHubDeviceFlow(
  pending: DeviceFlowPending,
  signal?: AbortSignal,
  onAttempt?: (attempt: number) => void,
): Promise<string> {
  const clientId = getGitHubClientId();
  if (!clientId) {
    throw new Error("GitHub client ID is not configured.");
  }

  if (!pending.deviceCode) {
    throw new Error("Missing device code. Please try signing in again.");
  }

  if (isTauri()) {
    return pollGitHubDeviceFlowTauri(pending, signal, onAttempt);
  }

  const deadline = Date.now() + pending.expiresIn * 1000;
  let intervalMs = pending.interval * 1000;
  let attempt = 0;

  while (Date.now() < deadline) {
    if (signal?.aborted) {
      throw new Error("Sign-in cancelled");
    }

    attempt += 1;
    onAttempt?.(attempt);

    const data = await postFormBrowser(`${OAUTH_BASE}${ACCESS_TOKEN_PATH}`, {
      client_id: clientId,
      device_code: pending.deviceCode,
      grant_type: "urn:ietf:params:oauth:grant-type:device_code",
    });

    if (data.access_token) {
      return data.access_token;
    }

    const error = data.error;
    if (error === "authorization_pending") {
      // fall through to sleep
    } else if (error === "slow_down") {
      intervalMs += 5000;
    } else if (error === "expired_token") {
      throw new Error("The sign-in code expired. Please try again.");
    } else if (error === "access_denied") {
      throw new Error("GitHub sign-in was denied.");
    } else {
      throw new Error(data.error_description ?? error ?? "GitHub sign-in failed");
    }

    await delay(intervalMs);
  }

  throw new Error("Sign-in timed out. Please try again.");
}

export async function fetchGitHubUser(accessToken: string): Promise<AuthUser> {
  if (isTauri()) {
    const user = await invoke<AuthUser & { avatarUrl?: string }>("github_fetch_user", {
      accessToken,
    });
    return normalizeAuthUser(user);
  }

  const res = await fetch(`${API_BASE}${USER_API_PATH}`, {
    headers: {
      Accept: "application/vnd.github+json",
      Authorization: `Bearer ${accessToken}`,
      "X-GitHub-Api-Version": "2022-11-28",
      "User-Agent": "FORGE/2.0",
    },
  });

  if (!res.ok) {
    throw new Error("Could not load your GitHub profile.");
  }

  const data = (await res.json()) as AuthUser;
  return normalizeAuthUser(data);
}

export async function completeGitHubDeviceLogin(
  pending: DeviceFlowPending,
  signal?: AbortSignal,
  onPollAttempt?: (attempt: number) => void,
): Promise<AuthSession> {
  const clientId = getGitHubClientId();
  if (!clientId) {
    throw new Error("GitHub client ID is not configured.");
  }

  if (!pending.deviceCode) {
    throw new Error("Missing device code. Please try signing in again.");
  }

  const accessToken = await pollGitHubDeviceFlow(pending, signal, onPollAttempt);
  const user = await fetchGitHubUser(accessToken);
  return normalizeAuthSession({
    accessToken,
    user,
    provider: "github",
    loginAt: Date.now(),
  });
}

File: src-tauri/src/github_auth.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

const DEVICE_CODE_URL: &str = "https://github.com/login/device/code";
const ACCESS_TOKEN_URL: &str = "https://github.com/login/oauth/access_token";
const USER_API_URL: &str = "https://api.github.com/user";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeviceFlowPending {
    pub device_code: String,
    pub user_code: String,
    pub verification_uri: String,
    pub expires_in: u64,
    pub interval: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitHubUser {
    pub id: u64,
    pub login: String,
    pub name: Option<String>,
    pub avatar_url: String,
    pub email: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AuthSession {
    pub access_token: String,
    pub user: GitHubUser,
    pub login_at: u64,
}

async fn post_form(
    url: &str,
    body: &HashMap<String, String>,
) -> Result<HashMap<String, serde_json::Value>, String> {
    let client = reqwest::Client::new();
    let res = client
        .post(url)
        .header("Accept", "application/json")
        .form(body)
        .send()
        .await
        .map_err(|e| format!("Network error contacting GitHub: {e}"))?;

    let status = res.status();
    let data: HashMap<String, serde_json::Value> = res
        .json()
        .await
        .map_err(|e| format!("Invalid response from GitHub: {e}"))?;

    if !status.is_success() {
        let message = data
            .get("error_description")
            .or_else(|| data.get("error"))
            .and_then(|v| v.as_str())
            .unwrap_or("GitHub request failed");
        return Err(message.to_string());
    }

    Ok(data)
}

fn value_as_string(map: &HashMap<String, serde_json::Value>, key: &str) -> Option<String> {
    map.get(key).and_then(|v| v.as_str()).map(str::to_string)
}

fn value_as_u64(map: &HashMap<String, serde_json::Value>, key: &str) -> Option<u64> {
    map.get(key).and_then(|v| v.as_u64())
}

async fn poll_device_token_once(
    client: &reqwest::Client,
    client_id: &str,
    device_code: &str,
) -> Result<HashMap<String, serde_json::Value>, String> {
    let mut body: HashMap<String, String> = HashMap::new();
    body.insert("client_id".to_string(), client_id.to_string());
    body.insert("device_code".to_string(), device_code.to_string());
    body.insert(
        "grant_type".to_string(),
        "urn:ietf:params:oauth:grant-type:device_code".to_string(),
    );

    let res = client
        .post(ACCESS_TOKEN_URL)
        .header("Accept", "application/json")
        .form(&body)
        .send()
        .await
        .map_err(|e| format!("Network error contacting GitHub: {e}"))?;

    res.json()
        .await
        .map_err(|e| format!("Invalid response from GitHub: {e}"))
}

enum PollOutcome {
    Token(String),
    Pending,
    SlowDown,
}

fn parse_token_poll_response(
    data: HashMap<String, serde_json::Value>,
) -> Result<PollOutcome, String> {
    if let Some(token) = value_as_string(&data, "access_token") {
        return Ok(PollOutcome::Token(token));
    }

    let error = value_as_string(&data, "error").unwrap_or_default();
    match error.as_str() {
        "authorization_pending" => Ok(PollOutcome::Pending),
        "slow_down" => Ok(PollOutcome::SlowDown),
        "expired_token" => Err("The sign-in code expired. Please try again.".into()),
        "access_denied" => Err("GitHub sign-in was denied.".into()),
        "" => {
            let message = data
                .get("error_description")
                .and_then(|v| v.as_str())
                .unwrap_or("GitHub sign-in failed");
            Err(message.to_string())
        }
        _ => {
            let message = data
                .get("error_description")
                .and_then(|v| v.as_str())
                .unwrap_or(&error);
            Err(message.to_string())
        }
    }
}

async fn wait_for_device_token(
    client_id: &str,
    device_code: &str,
    interval_secs: u64,
    expires_in_secs: u64,
) -> Result<String, String> {
    let client_id = client_id.trim();
    let device_code = device_code.trim();
    if client_id.is_empty() {
        return Err("GitHub client ID is not configured.".into());
    }
    if device_code.is_empty() {
        return Err("Missing device code. Please try signing in again.".into());
    }

    let http = reqwest::Client::new();
    let deadline = Instant::now() + Duration::from_secs(expires_in_secs.max(60));
    let mut interval = Duration::from_secs(interval_secs.max(2));

    loop {
        if Instant::now() >= deadline {
            return Err("Sign-in timed out. Please try again.".into());
        }

        let data = poll_device_token_once(&http, client_id, device_code).await?;

        match parse_token_poll_response(data)? {
            PollOutcome::Token(token) => return Ok(token),
            PollOutcome::SlowDown => interval += Duration::from_secs(5),
            PollOutcome::Pending => {}
        }

        tokio::time::sleep(interval).await;
    }
}

async fn fetch_github_user(access_token: &str) -> Result<GitHubUser, String> {
    let access_token = access_token.trim();
    if access_token.is_empty() {
        return Err("Missing access token.".into());
    }

    let client = reqwest::Client::new();
    let res = client
        .get(USER_API_URL)
        .header("Accept", "application/vnd.github+json")
        .header("Authorization", format!("Bearer {access_token}"))
        .header("X-GitHub-Api-Version", "2022-11-28")
        .send()
        .await
        .map_err(|e| format!("Network error contacting GitHub: {e}"))?;

    if !res.status().is_success() {
        return Err("Could not load your GitHub profile.".into());
    }

    let data: GitHubUser = res
        .json()
        .await
        .map_err(|e| format!("Invalid profile response from GitHub: {e}"))?;

    Ok(data)
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

#[tauri::command]
pub async fn github_start_device_flow(
    client_id: String,
    scope: String,
) -> Result<DeviceFlowPending, String> {
    let client_id = client_id.trim().to_string();
    if client_id.is_empty() {
        return Err("GitHub client ID is not configured.".into());
    }

    let mut body: HashMap<String, String> = HashMap::new();
    body.insert("client_id".to_string(), client_id);
    body.insert("scope".to_string(), scope);

    let data = post_form(DEVICE_CODE_URL, &body).await?;

    let device_code = value_as_string(&data, "device_code")
        .ok_or("GitHub did not return a device code.")?;
    let user_code = value_as_string(&data, "user_code")
        .ok_or("GitHub did not return a user code.")?;
    let verification_uri = value_as_string(&data, "verification_uri")
        .ok_or("GitHub did not return a verification URL.")?;

    Ok(DeviceFlowPending {
        device_code,
        user_code,
        verification_uri,
        expires_in: value_as_u64(&data, "expires_in").unwrap_or(900),
        interval: value_as_u64(&data, "interval").unwrap_or(5),
    })
}

#[tauri::command]
pub async fn github_wait_for_device_token(
    client_id: String,
    device_code: String,
    interval_secs: u64,
    expires_in_secs: u64,
) -> Result<String, String> {
    wait_for_device_token(
        &client_id,
        &device_code,
        interval_secs,
        expires_in_secs,
    )
    .await
}

/// Wait for GitHub authorization, fetch the user profile, and return a session.
#[tauri::command]
pub async fn github_complete_device_login(
    client_id: String,
    device_code: String,
    interval_secs: u64,
    expires_in_secs: u64,
) -> Result<AuthSession, String> {
    let access_token = wait_for_device_token(
        &client_id,
        &device_code,
        interval_secs,
        expires_in_secs,
    )
    .await?;
    let user = fetch_github_user(&access_token).await?;

    Ok(AuthSession {
        access_token,
        user,
        login_at: now_ms(),
    })
}

#[tauri::command]
pub async fn github_fetch_user(access_token: String) -> Result<GitHubUser, String> {
    fetch_github_user(&access_token).await
}
